### ISSUE TITLE
Style Transfer - Model Architecture

### ISSUE THREAD

**Issue by prash-p** [apple/turicreate#1128]
Hi, 
I wanted to find out how the style transfer architecture is set up in turicreate. In the method published by Gatys (https://arxiv.org/abs/1508.06576), the style and content is optimized given two unseen style and content images and a pretrained network (e.g. vgg-19 imagenet). What is the turicreate model learning from the training set of style and content images?

**Comment by abhishekpratapa** [comment]
Thanks Prash for the question. Here is the specific implementation we use in the current style transfer toolkit: https://arxiv.org/pdf/1607.08022.pdf

## Networks

There are two different networks used.

- A generator network we call Transformer. `src/unity/python/turicreate/toolkits/style_transfer/_model.py:106`

- A VGG-16 network that's used for training the Transformer Network. `src/unity/python/turicreate/toolkits/style_transfer/_model.py:192`

## Training

The training process is as such.

### Forward Pass
- A forward pass through the Transformer Network using the Content Images.
- A forward pass through the VGG-16 Network using both the Content and Style Images respectively

### Style Loss
- We keep track of the outputs after some of the activation layers `[relu1_2, relu2_2, relu3_3, relu4_3]` we calculate the gram matrix using the outputs of these activation layers for the Content and Style Images respectively. `src/unity/python/turicreate/toolkits/style_transfer/style_transfer.py:316`

- The addition of these losses is the `style_loss`

### Content Loss
- An additional L2 loss is computed between the output `[relu3_3]` after the forward pass of the Content and Style Images we call this the `content_loss`

### Total Loss
- We add the losses together and only update the beta and gamma weights in the InstanceNorm layers of the Transformer Network.

Repeating this training process in a training loop allows us to create a Transformer network that can quickly stylize images. See `src/unity/python/turicreate/toolkits/style_transfer/style_transfer.py:283` for the training loop.

I hope this explained the process in which we train the style transfer model.
